app/core/models/ocelot/LabelCache.scala

```diff
@@ -85,24 +86,33 @@ private class LabelCacheImpl(labels: Map[String, Label] = Map(),
     cache + (name -> cache.get(name).fold[Label](ListLabel(name, english, welsh))(l => ListLabel(l.name, english, welsh)))
 
   // Flows
-  def pushFlows(flowNext: Seq[String], continue: String, labelName: Option[String], labelValues: Seq[String], stanzas: Map[String, Stanza]): Labels =
+  def pushFlows(flowNext: Seq[String], continue: String, labelName: Option[String], labelValues: Seq[String], stanzas: Map[String, Stanza]): (Option[String], Labels) =
     flowNext.zipWithIndex.map{
-      case (nxt, idx) => Flow(nxt, labelName.map(LabelValue(_, labelValues.lift(idx).fold[Option[String]](None)(v => Some(v)))))
+      case (nxt, idx) => Flow(nxt, labelName.map(LabelValue(_, labelValues(idx))))
     } match {
-      case Nil => this
-      case flows => new LabelCacheImpl(labels, cache, flows.toList ++ (Continuation(continue) :: stack), pool, poolCache ++ stanzas)
+      case Nil => (Some(continue), this)
+      case x :: xs =>
+        (Some(x.next),
+         x.labelValue.fold(new LabelCacheImpl(labels, cache, x :: xs ++ (Continuation(continue) :: stack), pool, poolCache ++ stanzas))
+                          (lv => new LabelCacheImpl(labels, updateOrAddScalarLabel(lv.name, lv.value, None), x :: xs ++ (Continuation(continue) :: stack), pool, poolCache ++ stanzas))
+        )
     }
 
-  def takeFlow: Option[(String, Labels)] = // Remove head of flow stack and update flow label if required
-    stack.headOption.map{
-      case f: Flow =>
-      (f.next,
-       f.labelValue.flatMap(lv => lv.value.map(v => new LabelCacheImpl(labels, updateOrAddScalarLabel(lv.name, v, None), stack.tail, pool, poolCache)))
-        .getOrElse(new LabelCacheImpl(labels, cache, stack.tail, pool, poolCache))
-      )
-      case c: Continuation => (c.next, new LabelCacheImpl(labels, cache, stack.tail, pool, poolCache))
+  def nextFlow: Option[(String, Labels)] = // Remove head of flow stack and update flow label if required
+    stack match {
```

Comment:
As I don't think x is used anywhere we could replace it with _. However, that might make things even more confusing.

Reply:
I've added this update to current ticket which affects the same area